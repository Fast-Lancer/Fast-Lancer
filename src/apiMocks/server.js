import { rest } from 'msw'
import { setupServer } from 'msw/node'

const url = process.env.REACT_APP_SUPABASE_URL + '/rest/v1'

let postedClient = {}
let postedProject = {}

export const server = setupServer(
  rest.get(url + '/clients', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(select === '*' && !id) {
      // response for getClients()
      return res(
        ctx.json([
          {
            id: 1,
            client_name: 'bob1',
            business_name: 'business1'
          },
          {
            id: 2,
            client_name: 'bob2',
            business_name: 'business2'
          },
          {
            id: 3,
            client_name: 'bob3',
            business_name: 'business3'
          }
        ])
      )
    } else if(id) {
      // response for getClient(id)
      if(id === 'eq.42') {
        // Janky workaround to make testing edit/create possible.
        // The post route mock will set the id to 42
        // So this responds with the posted data when client.id is 42
        return res(
          ctx.json(
            postedClient
          )
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              client_name: 'bob1',
              business_name: 'business1'
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/clients', (req, res, ctx) => {
    let [client] = req.body
    client = { ...client, id: 42 }
    postedClient = client
    return res(
      ctx.json([
        client
      ])
    )
  }),
  rest.get(url + '/projects', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(!id) {
      // response for getProjects()
      return res(
        ctx.json([
          {
            id: 1,
            title: 'project1',
            client_id: '1',
            clients: {
              client_name: 'bob1'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          },
          {
            id: 2,
            title: 'project2',
            client_id: '2',
            clients: {
              client_name: 'bob2'
            },
            
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          
          },
          {
            id: 3,
            title: 'project3',
            client_id: '3',
            clients: {
              client_name: 'bob3'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          }
        ])
      )
    } else if(id) {
      // response for getProjectById(id)
      if(id === 'eq.42') {
        return res(
          ctx.json(postedProject)
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              title: 'project1',
              client_id: '1',
              clients: {
                client_name: 'bob1'
              },
              date_start:'2022-01-20',
              date_end:'2022-01-21',
              URL: 'something@something.com',
              description: 'ok',
              notes: 'test notes',
              hourly_rate: 33,
              hours_quoted: 22,
              price_quoted: 300
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/projects', (req, res, ctx) => {
    let project = req.body
    project = { ...project, id: 42 }
    postedProject = project
    return res(
      ctx.json([project])
    )
  })
)
