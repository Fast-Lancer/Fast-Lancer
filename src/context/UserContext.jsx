import { createContext, useContext, useMemo, useState } from 'react'
import { getUser } from '../services/users'

const UserContext = createContext({})

export function UserProvider({ children }) {
  const currentUser = getUser()
  const [user, setUser] = useState(currentUser ?? {})
  const memo = useMemo(() => ({ user, setUser }), [user])
  return (
    <UserContext.Provider value={memo}>
      {children}
    </UserContext.Provider>
  )
}

export function useUser() {
  const context = useContext(UserContext)

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider')
  }

  return context
};
