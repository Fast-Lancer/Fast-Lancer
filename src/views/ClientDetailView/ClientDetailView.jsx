import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import ClientDetail from '../../components/ClientDetail/ClientDetail.jsx'
import Title from '../../components/Title/Title.jsx'
import { getClient } from '../../services/clients.js'

export default function ClientDetailView() {
  const { id } = useParams()
  const [client, setClient] = useState()
  const [loading, setLoading] = useState(true)

  useEffect(() => {
    getClient(id)
      .then(setClient)
      .catch(e => console.error(e))
      .finally(() => setLoading(false))
  }, [])
  
  return (
    <div>
      {
        loading
          ? <h1>Loading...</h1>
          : <div>
            <Title pageTitle='client detail' pageHeader={client.name} />
            <ClientDetail client={client}/>
          </div>
      }
    </div>
  )
}
