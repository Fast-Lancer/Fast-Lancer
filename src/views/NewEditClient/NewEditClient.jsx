import { useEffect, useState } from 'react'
import { useHistory, useParams } from 'react-router-dom'
import { updateClient, createClient, getClients } from '../../services/clients'
import Title from '../../components/Title/Title'
import NewEditForm from '../../components/NewEditForm/NewEditForm'


export default function NewEditClient({ isNew = false }) {
  // const [clients, setClients] = useState([])
  const history = useHistory()


  useEffect(() => {
    const clients = getClients()
    setClients(clients)
  }, [])

  let client

  const formSubmit = async ({ client_name, email, phone, business_name, notes, user_id }) => {
    try {
      if (isNew) {
        client = await createClient({ client_name, email, phone, business_name, notes, user_id })
        history.push(`/clients/${client[0].id}`)
      } else {
        client = await updateClient({ client_name, email, phone, business_name, notes, user_id })
        history.push(`/clients/${client.id}`)
      }
    } catch (error) {
      throw error
    }
  }

  return <div>
    <Title pageTitle='new-edit client' pageHeader='New-Edit Client'/>
    <main>
      <NewEditForm 
        formSubmit={formSubmit}
        formLabel={ isNew ? 'Add New Client' : 'Update Client' }
        // editPlaceholders={ isNew ? null : { ...clients }} 
      />
    </main>
  </div>
}
