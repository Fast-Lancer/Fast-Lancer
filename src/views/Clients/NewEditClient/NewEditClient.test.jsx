import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { useEffect } from 'react'
import { Route, Switch } from 'react-router-dom'
import { useLocation } from 'react-router-dom'
import { MemoryRouter } from 'react-router-dom'
import { server } from '../../../apiMocks/server.js'
import { UserProvider, useUser } from '../../../context/UserContext.jsx'
import ClientDetailView from '../ClientDetailView/ClientDetailView.jsx'
import NewEditClient from './NewEditClient.jsx'

beforeAll(() => {
  server.listen()
})

afterAll(() => {
  server.close()
})

const UserAdder = () => {
  const { setUser } = useUser()
  useEffect(() => {
    setUser({ id: 1 })
  }, [])

  return null
}

let location

const LocationGrabber = () => {
  location = useLocation()
  return null
}

it.skip('should render the fetched data', async () => {
  const { container } = render(
    <MemoryRouter initialEntries={['/']}>
      <UserProvider>
        <UserAdder />
        <Switch>
          <Route path='/clients/:id'>
            <LocationGrabber />
            <ClientDetailView />
          </Route>
          <Route path='/'>
            <LocationGrabber />
            <NewEditClient isNew />
          </Route>
        </Switch>
      </UserProvider>
    </MemoryRouter>
  )

  const nameInput = await screen.findByLabelText('Client Name')
  const emailInput = await screen.findByLabelText('E-Mail')
  const phoneInput = await screen.findByLabelText('Phone Number')
  const businessInput = await screen.findByLabelText('Business Name')
  const notesInput = await screen.findByLabelText('Notes')
  const submitButton = await screen.findByText('Save')

  fireEvent.change(nameInput, { target: { value: 'bobs name' } })
  fireEvent.change(emailInput, { target: { value: 'bob@bob.com' } })
  fireEvent.change(phoneInput, { target: { value: '123' } })
  fireEvent.change(businessInput, { target: { value: 'bobs business' } })
  fireEvent.change(notesInput, { target: { value: 'note' } })

  fireEvent.click(submitButton)

  await waitFor(() => location.pathName === '/clients/42', { timeout: 1000 })

  await screen.findByText('bobs business')

  expect(container).toMatchSnapshot()
})
