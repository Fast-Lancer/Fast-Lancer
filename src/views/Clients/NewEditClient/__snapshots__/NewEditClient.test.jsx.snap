// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the fetched data 1`] = `
<div>
  <div>
    <div>
      <div
        class="titleBody"
      >
        <section
          class="title"
        >
          bobs name
        </section>
        <ul
          class="tabnav"
        >
          <li
            class="active"
          >
            <a
              href="/clients"
            >
              Clients
            </a>
          </li>
          <li>
            <a
              href="/projects"
            >
              Projects
            </a>
          </li>
          <li>
            <a
              href="/clients/edit/42"
            >
              Edit Client
            </a>
          </li>
        </ul>
      </div>
      <main>
        <div
          class="clientDetail"
        >
          <h3
            class="h3"
          >
            bobs business
          </h3>
          <div
            class="twoColRow"
          >
            <div
              class="col"
            >
              <div
                class="row"
              >
                <img
                  alt="a phone icon"
                  src="phoneSmall.png"
                />
                123
              </div>
              <div
                class="row"
              >
                <img
                  alt="an email icon"
                  src="emailSmall.png"
                />
                bob@bob.com
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="row"
              >
                <img
                  alt="a palette icon"
                  src="paletteSmall.png"
                />
                Client Projects
              </div>
              <ul
                class="linkList"
              >
                <li>
                  <a
                    href="/projects/1"
                  >
                    project1
                  </a>
                </li>
                <li>
                  <a
                    href="/projects/2"
                  >
                    project2
                  </a>
                </li>
                <li>
                  <a
                    href="/projects/3"
                  >
                    project3
                  </a>
                </li>
              </ul>
            </div>
          </div>
          <div>
            <div
              class="row"
            >
              <img
                alt="a notes icon"
                src="emailSmall.png"
              />
              Notes
            </div>
            <div
              class="notesDiv"
            >
              note
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
